File: src/main/java/com/example/myprog/java/examples/immutability/ImmutableObject.java
package com.example.myprog.java.examples.immutability;

import java.util.ArrayList;
import java.util.List;

public class ImmutableObject {

    private final String name;
    private final List<Integer> numbers;

    public ImmutableObject(String name, List<Integer> numbers) {
        this.name = name;
        this.numbers = numbers;
    }

    public String getName() {
        return name;
    }

    // Exposing this would allow callers to modify the List!
//    public List<Integer> getNumbers() {
//        return numbers;
//    }

    // better to only expose what we need, for example:

    public Integer getNumber(int index) {
        return numbers.get(index);
    }

    public int getNumbersSize() {
        return numbers.size();
    }

}

class ImmutableObjectRunner {
    public static void main(String[] args) {
        var list = new ArrayList<Integer>();
        list.add(1);
        list.add(2);
        list.add(3);
        var a = new ImmutableObject("someone", list);

        a.getNumber(1);
    }
}

File: src/main/kotlin/com/example/myprog/kotlin/examples/immutability/ImmutableObject.kt
package com.example.myprog.kotlin.examples.immutability

class ImmutableObject(val name: String, numbers: List<Int>) {

    /*
     * Let's say we want to modify 'numbers' internally in several places,
     * it's more convenient to save it as a MutableList.
     */
    private val numbers: MutableList<Int> = numbers.toMutableList()

    /* Then we can decide to expose only the 'List' interface so callers cannot modify the list */
    fun getNumbers(): List<Int> {
        return numbers
    }
}
